external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,517 @@
+                    kafkaSpoutConfig.getConsumerGroupId(), kafkaConsumer, partitions);
+
+            initialize(partitions);
+        }
+
+        private void initialize(Collection<TopicPartition> partitions) {
+            if (!consumerAutoCommitMode) {
+                acked.keySet().retainAll(partitions);   // remove from acked all partitions that are no longer assigned to this spout
+            }
+
+            for (TopicPartition tp : partitions) {
+                final OffsetAndMetadata committedOffset = kafkaConsumer.committed(tp);
+                final long fetchOffset = doSeek(tp, committedOffset);
+                setAcked(tp, fetchOffset);
+            }
+            initialized = true;
+            LOG.debug("Initialization complete");
+        }
+
+        /**
+         * sets the cursor to the location dictated by the first poll strategy and returns the fetch offset
+         */
+        private long doSeek(TopicPartition tp, OffsetAndMetadata committedOffset) {
+            long fetchOffset;
+            if (committedOffset != null) {             // offset was committed for this TopicPartition
+                if (firstPollOffsetStrategy.equals(EARLIEST)) {
+                    kafkaConsumer.seekToBeginning(tp);
+                    fetchOffset = kafkaConsumer.position(tp);
+                } else if (firstPollOffsetStrategy.equals(LATEST)) {
+                    kafkaConsumer.seekToEnd(tp);
+                    fetchOffset = kafkaConsumer.position(tp);
+                } else {
+                    // By default polling starts at the last committed offset. +1 to point fetch to the first uncommitted offset.
+                    fetchOffset = committedOffset.offset() + 1;
+                    kafkaConsumer.seek(tp, fetchOffset);
+                }
+            } else {    // no commits have ever been done, so start at the beginning or end depending on the strategy
+                if (firstPollOffsetStrategy.equals(EARLIEST) || firstPollOffsetStrategy.equals(UNCOMMITTED_EARLIEST)) {
+                    kafkaConsumer.seekToBeginning(tp);
+                } else if (firstPollOffsetStrategy.equals(LATEST) || firstPollOffsetStrategy.equals(UNCOMMITTED_LATEST)) {
+                    kafkaConsumer.seekToEnd(tp);
+                }
+                fetchOffset = kafkaConsumer.position(tp);
+            }
+            return fetchOffset;
+        }
+    }
+
+    private void setAcked(TopicPartition tp, long fetchOffset) {
+        // If this partition was previously assigned to this spout, leave the acked offsets as they were to resume where it left off
+        if (!consumerAutoCommitMode && !acked.containsKey(tp)) {
+            acked.put(tp, new OffsetEntry(tp, fetchOffset));
+        }
+    }
+
+    // ======== Next Tuple =======
+
+    @Override
+    public void nextTuple() {
+        if (initialized) {
+            if (commit()) {
+                commitOffsetsForAckedTuples();
+            } else if (poll()) {
```

Comment:
Minor nit. I don't think poll, commit, and logging should be mutually exclusive.  It slows things down if we don't emit anything, and it also means that we may never log if we always emit something.

```
if (commit()) {
    commitOffsetsForAckedTuples();
}

if (poll()) {
  emitTuples(pollKafkaBroker());
}

if (logTimer.isExpiredResetOnTrue()) {
  log()
}
```